lib/p5.play.js

```diff
@@ -2189,246 +2123,230 @@ function Sprite(pInst, _x, _y, _w, _h) {
-                  }
+    if (type === 'collide' || type === 'displace' || type === 'bounce') {
+      if (type === 'displace' && !target.immovable) {
+        target.position.sub(displacement);
+      } else if ((type === 'collide' || type === 'bounce') && !this.immovable) {
+        this.position.add(displacement);
+        this.previousPosition = createVector(this.position.x, this.position.y);
+        this.newPosition = createVector(this.position.x, this.position.y);
+        this.collider.updateFromSprite(this);
+      }
 
-                if(tunnelY) {
-                  //from top
-                  if(this.velocity.y > 0)
-                    displacement.y = other.collider.top() - this.collider.bottom() - 1;
-                  else if(this.velocity.y < 0 )
-                    displacement.y = other.collider.bottom() - this.collider.top() + 1;
+      if (displacement.x > 0)
+        this.touching.left = true;
+      if (displacement.x < 0)
+        this.touching.right = true;
+      if (displacement.y < 0)
+        this.touching.bottom = true;
+      if (displacement.y > 0)
+        this.touching.top = true;
+
+      if (type === 'bounce') {
+        var thisOldVelOnAxis = p5.Vector.project(this.velocity, displacement);
+        var targetOldVelOnAxis = p5.Vector.project(target.velocity, displacement);
+        var dx1 = displacement.copy();
+        var dx2 = p5.Vector.mult(dx1, -1);
+        var magnitude = dx1.magSq();
+        var totalMass = this.mass + target.mass;
+        var m1 = 0, m2 = 0;
+        if (this.immovable) {
+          m2 = 1;
+        } else if (target.immovable) {
+          m1 = 1;
+        } else {
+          m1 = target.mass / totalMass;
+          m2 = this.mass / totalMass;
+        }
+        var newVel1 = dx1.mult(m1 * p5.Vector.sub(this.velocity, target.velocity).dot(dx1) / magnitude);
+        var newVel2 = dx2.mult(m2 * p5.Vector.sub(target.velocity, this.velocity).dot(dx2) / magnitude);
 
-                  }
+        this.velocity.sub(thisOldVelOnAxis).sub(newVel1.mult(this.restitution));
+        target.velocity.sub(targetOldVelOnAxis).sub(newVel2.mult(target.restitution));
+      }
+    }
 
-              }//end overlap
+    // Finally, for all collision types call the callback and record
+    // that collision occurred.
+    if (typeof callback === 'function') {
+      callback.call(this, this, target);
+    }
+    return true;
+  };
 
+  this._findDisplacement = function(target) {
+    // Multisample if tunneling occurs:
+    // If the direction between the two
+    // If the relative velocities of the two sprites is greater than the
+    // difference of their previous centers on the velocity axis,
+    // their centers might have crossed on an axis this frame which would permit
+    // tunneling.
+    // In that case, test interpolations between their last positions and their
+    // current positions, and check for tunneling that way.
```

Comment:
Feedback here too, please.  We prevent tunneling with a multisampling approach, which seems to work really well and perform okay.  I did some profiling; in typical cases we do 2-3 sub-frame samples and manage to compute a displacement within a millisecond.  After I optimized a bit even in really bad cases where we take nearly 100 sub-frame samples you still get a displacement in 20-25ms.  I haven't put an upper bound on the number of sub-frame samples we'll take - it approaches an asymptote as you get into extreme cases which are likely to be rare anyway, but I think it's worth it to handle things nicely at any speed.

![tunneling_one_frame](https://cloud.githubusercontent.com/assets/1615761/20023906/37e70c18-a2a0-11e6-9ce5-2a030856cdd9.png)